rust/src/bls/api.rs

```diff
@@ -55,105 +57,88 @@ macro_rules! try_ffi {
 ///
 /// # Arguments
 ///
-/// * `message_ptr` - pointer to a message byte array
-/// * `message_len` - length of the byte array
-#[no_mangle]
-pub unsafe extern "C" fn fil_hash(
-    message_ptr: *const u8,
-    message_len: libc::size_t,
-) -> *mut types::fil_HashResponse {
-    // prep request
-    let message = from_raw_parts(message_ptr, message_len);
-
+/// * `message` - reference to a message byte array
+#[ffi_export]
+pub fn hash(message: c_slice::Ref<u8>) -> repr_c::Box<types::HashResponse> {
     // call method
-    let digest = hash_sig(message);
+    let digest = hash_sig(&message);
 
     // prep response
     let mut raw_digest: [u8; DIGEST_BYTES] = [0; DIGEST_BYTES];
     raw_digest.copy_from_slice(digest.to_bytes().as_ref());
 
-    let response = types::fil_HashResponse {
-        digest: fil_BLSDigest { inner: raw_digest },
+    let response = types::HashResponse {
+        digest: BLSDigest { inner: raw_digest },
     };
 
-    Box::into_raw(Box::new(response))
+    repr_c::Box::new(response)
 }
 
 /// Aggregate signatures together into a new signature
 ///
 /// # Arguments
 ///
-/// * `flattened_signatures_ptr` - pointer to a byte array containing signatures
-/// * `flattened_signatures_len` - length of the byte array (multiple of SIGNATURE_BYTES)
+/// * `flattened_signatures` - byte array containing signatures
 ///
 /// Returns `NULL` on error. Result must be freed using `destroy_aggregate_response`.
-#[no_mangle]
-pub unsafe extern "C" fn fil_aggregate(
-    flattened_signatures_ptr: *const u8,
-    flattened_signatures_len: libc::size_t,
-) -> *mut types::fil_AggregateResponse {
+#[ffi_export]
+pub fn aggregate(
+    flattened_signatures: c_slice::Ref<u8>,
+) -> Option<repr_c::Box<types::AggregateResponse>> {
     // prep request
     let signatures = try_ffi!(
-        from_raw_parts(flattened_signatures_ptr, flattened_signatures_len)
+        flattened_signatures
             .par_chunks(SIGNATURE_BYTES)
             .map(|item| { Signature::from_bytes(item) })
             .collect::<Result<Vec<_>, _>>(),
-        std::ptr::null_mut()
+        None
     );
 
     let mut raw_signature: [u8; SIGNATURE_BYTES] = [0; SIGNATURE_BYTES];
 
-    let aggregated = try_ffi!(aggregate_sig(&signatures), std::ptr::null_mut());
+    let aggregated = try_ffi!(aggregate_sig(&signatures), None);
     aggregated
         .write_bytes(&mut raw_signature.as_mut())
         .expect("preallocated");
 
-    let response = types::fil_AggregateResponse {
-        signature: fil_BLSSignature {
+    let response = types::AggregateResponse {
+        signature: BLSSignature {
             inner: raw_signature,
         },
     };
```

Comment:
Just an idea, I don't know if that's Rustic or not (or a good idea). Could this perhaps be something like:

```rust
let signature = BLSSignature::new();
aggregated
    .write_bytes(&mut signature.as_mut())
    .expect("preallocated");
```

The same would then apply to the other BLS things.

Reply:
Wow that's really nice and much shorter now.